examples/credential-registry/src/lib.rs

```diff
@@ -0,0 +1,1292 @@
+    }
+
+    fn register_credential(
+        &mut self,
+        credential_id: Uuidv4,
+        credential_data: &CredentialData,
+    ) -> ContractResult<()> {
+        let credential_entry = CredentialEntry {
+            credential_data:  credential_data.clone(),
+            revocation_nonce: 0,
+            is_revoked:       false,
+        };
+        let res = self.credentials.insert(credential_id, credential_entry);
+        ensure!(res.is_none(), ContractError::CredentialAlreadyExists);
+        Ok(())
+    }
+
+    fn update_credential(
+        &mut self,
+        credential_id: Uuidv4,
+        credential_data: &CredentialData,
+    ) -> ContractResult<()> {
+        let mut registry_entry =
+            self.credentials.entry(credential_id).occupied_or(ContractError::CredentialNotFound)?;
+        registry_entry.credential_data = credential_data.clone();
+        Ok(())
+    }
+
+    fn revoke_credential(&mut self, now: Timestamp, credential_id: Uuidv4) -> ContractResult<()> {
+        let mut credential =
+            self.credentials.get_mut(&credential_id).ok_or(ContractError::CredentialNotFound)?;
+        let status = credential.get_status(now);
+        ensure!(
+            status == CredentialStatus::Active || status == CredentialStatus::NotActivated,
+            ContractError::IncorrectStatusBeforeRevocation
+        );
+        credential.is_revoked = true;
+        Ok(())
+    }
+
+    fn register_issuer_key(&mut self, key_index: u8, pk: PublicKeyEd25519) -> ContractResult<()> {
+        let res = self.issuer_keys.insert(key_index, pk);
+        ensure!(res.is_none(), ContractError::KeyAlreadyExists);
+        Ok(())
+    }
+
+    fn remove_issuer_key(&mut self, key_index: u8) -> ContractResult<()> {
+        self.issuer_keys.remove(&key_index);
+        Ok(())
+    }
+
+    fn register_revocation_key(
+        &mut self,
+        key_index: u8,
+        pk: PublicKeyEd25519,
+    ) -> ContractResult<()> {
+        let res = self.revocation_keys.insert(key_index, (pk, 0));
+        ensure!(res.is_none(), ContractError::KeyAlreadyExists);
+        Ok(())
+    }
+
+    fn view_revocation_key(&self, key_index: u8) -> ContractResult<(PublicKeyEd25519, u64)> {
+        self.revocation_keys.get(&key_index).map(|x| *x).ok_or(ContractError::KeyNotFound)
+    }
+
+    fn remove_revocation_key(&mut self, key_index: u8) -> ContractResult<()> {
+        self.issuer_keys.remove(&key_index);
+        Ok(())
+    }
+
+    fn view_issuer_keys(&self) -> Vec<(u8, PublicKeyEd25519)> {
+        self.issuer_keys.iter().map(|x| (*x.0, *x.1)).collect()
+    }
+
+    fn view_issuer_metadata(&self) -> MetadataUrl { self.issuer_metadata.clone() }
+
+    fn update_issuer_metadata(&mut self, issuer_metadata: &MetadataUrl) {
+        self.issuer_metadata = issuer_metadata.clone()
+    }
+}
+
+/// Init function that creates a fresh registry state given the issuer's
+/// metadata
+#[init(contract = "credential_registry")]
```

Comment:
If you add an enum with the events it could be included in the schema by adding:
```suggestion
#[init(contract = "credential_registry", event = "Event")]
```